File: frontend/lib/activeEdit.ts
// Active Edit 클라이언트 미러 — spec §7.5. 서버 verify_active_edit(§13.1)와 동일 규칙.
// 키 입력마다 로컬 즉시 계산(서버 왕복 0). 제출 시 서버가 권위 재검증.

const norm = (t: string) => (t ?? "").normalize("NFKC").trim().replace(/\s+/g, " ");
const tok = (t: string) =>
  norm(t)
    .replace(/[^\p{L}\p{N}\s]/gu, " ")
    .split(/\s+/)
    .filter(Boolean);

function changedWords(o: string, m: string) {
  const a = tok(o),
    b = tok(m),
    n = a.length,
    M = b.length;
  const dp = Array.from({ length: n + 1 }, () => new Int32Array(M + 1));
  for (let i = 1; i <= n; i++)
    for (let j = 1; j <= M; j++)
      dp[i][j] = a[i - 1] === b[j - 1] ? dp[i - 1][j - 1] + 1 : Math.max(dp[i - 1][j], dp[i][j - 1]);
  const lcs = dp[n][M];
  return { changed: n - lcs + (M - lcs), ratio: n ? (n - lcs + (M - lcs)) / n : 0, modLen: M };
}

export type Tier = "none" | "trivial" | "ok" | "good";

export function evaluateEdit(
  orig: string,
  draft: string,
  { minWords = 1, suspicious = 0.05, good = 0.3 } = {}
) {
  const { changed, ratio, modLen } = changedWords(orig, draft);
  const identical = norm(orig) === norm(draft),
    empty = modLen === 0,
    valid = !identical && !empty && changed >= minWords;
  let tier: Tier = "none";
  if (valid) tier = ratio >= good ? "good" : ratio < suspicious ? "trivial" : "ok";
  return { valid, changed, ratio, tier, empty };
}

File: backend/app/services/active_edit.py
"""Active Edit 검증 — spec §13.1 그대로. 제출 시 서버 권위 검증."""
import re
import unicodedata
from difflib import SequenceMatcher

from app.core.config import settings


def _norm(t: str) -> str:
    return re.sub(r"\s+", " ", unicodedata.normalize("NFKC", t or "").strip())


def _tok(t: str) -> list[str]:
    return [w for w in re.sub(r"[^\w가-힣\s]", " ", _norm(t)).split() if w]


def diff_stats(orig: str, mod: str) -> dict:
    o, m = _tok(orig), _tok(mod)
    sm = SequenceMatcher(a=o, b=m, autojunk=False)
    rep = ins = dele = 0
    for tag, i1, i2, j1, j2 in sm.get_opcodes():
        if tag == "replace":
            rep += max(i2 - i1, j2 - j1)
        elif tag == "insert":
            ins += j2 - j1
        elif tag == "delete":
            dele += i2 - i1
    ch = rep + ins + dele
    return {
        "changed_words": ch,
        "change_ratio": ch / max(len(o), 1),
        "identical": _norm(orig) == _norm(mod),
        "mod_words": len(m),
    }


class ActiveEditError(Exception):
    ...


def verify_active_edit(original_a: str, modified_a: str) -> dict:
    s = diff_stats(original_a, modified_a)
    if s["mod_words"] == 0:
        raise ActiveEditError("정답이 비어 있습니다.")
    if s["identical"] or s["changed_words"] < settings.MIN_WORD_CHANGES:
        raise ActiveEditError("정답을 최소 1단어 이상 수정(패러프레이징)해야 합니다.")
    s["suspicious"] = s["change_ratio"] < settings.SUSPICIOUS_RATIO  # 차단X, 관리자 검토
    return s
